[WPE][GTK] Enable permissions request API

philn opened the pull request:
#### 80fe491a82749b5b90733f08cae3f2eac7286f9b
<pre>
[WPE][GTK] Enable permissions request API
<a href="https://bugs.webkit.org/show_bug.cgi?id=247065">https://bugs.webkit.org/show_bug.cgi?id=247065</a>

Reviewed by Carlos Garcia Campos.

This is exposed in the GLib API as a WebKitPermissionStateQuery signalled by the WebView using the
`query-permission-state` GObject signal. When the application has a handler for this signal it has
to eventually call `webkit_permission_state_query_finish()` with a valid `WebKitPermissionState`. If
not handled, the default result is `WEBKIT_PERMISSION_STATE_PROMPT`, thus instructing the User-Agent
that it has to explicitely present a prompt asking for permission to the given feature.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitAutocleanups.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitPermissionStateQuery.cpp: Added.
(_WebKitPermissionStateQuery::_WebKitPermissionStateQuery):
(_WebKitPermissionStateQuery::~_WebKitPermissionStateQuery):
(webkitPermissionStateQueryCreate):
(webkit_permission_state_query_ref):
(webkit_permission_state_query_unref):
(webkit_permission_state_query_get_name):
(webkit_permission_state_query_get_security_origin):
(webkit_permission_state_query_finish):
* Source/WebKit/UIProcess/API/glib/WebKitPermissionStateQuery.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitPermissionStateQueryPrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkit_web_view_class_init):
(webkitWebViewPermissionStateQuery):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/gtk/webkit.h:
* Source/WebKit/UIProcess/API/gtk/webkit2.h:
* Source/WebKit/UIProcess/API/wpe/webkit.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
(testWebViewQueryPermissionRequests):
(beforeAll):

Canonical link: <a href="https://commits.webkit.org/257787@main">https://commits.webkit.org/257787@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/9fc5ecd09f9e5d279987e9214bde2fec8c3c2ad5

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/100018 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/9189 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/33099 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/109365 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/169599 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/104017 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/10069 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/86543 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/92465 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/107264 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/105786 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/7608 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/90893 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/34326 "Passed tests") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/89515 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/22286 "Passed tests") | [⏳ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/API-Tests-GTK-EWS "Waiting to run tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/90625 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/2981 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/23801 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/86561 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/83/builds/388 "Built successfully and passed tests") | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/2940 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/60/builds/46173 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/29025 "Passed tests") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/9074 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/43282 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/89440 "Built successfully") | 
| [❌ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/5343 "Failed to push commit to Webkit repository") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/4790 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/19997 "Passed tests") | 
| | | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,110 @@
+ * WebKitQueryPermissionRequest:
+ * @See_also: #WebKitWebView
+ *
+ * This permission request represent a user's choice to allow or deny access to "powerful features"
```
This is not a permission request, we are not asking for permission, but the state of the permission request for different features. Maybe the request name here is confused and we could simply use `WebKitPermissionQuery` instead. `WebKitQueryPermissionRequest` works too, but the documentation should be clearer if we keep that name.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,110 @@
+ *
+ * When signalled by the #WebKitWebView through the `query-permission` signal, the application has
+ * to respond via return value of the signal, whether it grants, denies or requests a dedicated
+ * permission prompt for the given query.
```
`#WebKitWebView::query-permission` signal. Using the signal return value makes it synchronous and we don't even need this object because name and origin could be signal parameters. But we want to make it posible to handle the query asynchronously, since browser might have the permission states stored in a database or file in disk.

philn replied:
So then we need API for `_deny()`, `_grant()`, `_prompt()`?

carlosgcampos replied:
No, because again we are not granting or denying anything, we are saying that the permission stored by the browser for a particular feature in a particular origin is denied, granted or unknown/undecided. We could use an enum as you have in this patch and simply add `webkit_permission_query_finish()` with the query result.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,110 @@
+    WTF::String permissionName;
+};
+
+WEBKIT_DEFINE_TYPE(WebKitQueryPermissionRequest, webkit_query_permission_request, G_TYPE_OBJECT)
```
Maybe this could be a boxed type, instead of a GObject.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,110 @@
+
+void webkitQueryPermissionRequestFinish(WebKitQueryPermissionRequest* request, WebKitQueryPermissionResult result)
+{
+    auto state = WebCore::PermissionState::Prompt;
```
Move this to prompt case.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,110 @@
+const gchar*
+webkit_query_permission_get_name(WebKitQueryPermissionRequest* request)
+{
+    return request->priv->permissionName.utf8().data();
```
This is returning a temporary. We could use a `CString` in the struct and here simply return `.data()`.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,110 @@
+const gchar*
+webkit_query_permission_get_origin(WebKitQueryPermissionRequest* request)
+{
+    return request->priv->origin.utf8().data();
```
This is returning a temporary too. I think it would be better to return a `WebKitSecurityOrigin` though

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.h.in:
```diff
@@ -0,0 +1,85 @@
+    WEBKIT_QUERY_PERMISSION_GRANTED = 0,
+    WEBKIT_QUERY_PERMISSION_DENIED = 1,
+    WEBKIT_QUERY_PERMISSION_PROMPT = 2,
+} WebKitQueryPermissionResult;
```
We don't need to assign the values when using the deault.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.h.in:
```diff
@@ -0,0 +1,85 @@
+typedef struct _WebKitQueryPermissionRequestPrivate WebKitQueryPermissionRequestPrivate;
+
+/**
+ * WebKitQueryPermissionResult:
```
`WebKitPermissionState`?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequestPrivate.h:
```diff
@@ -0,0 +1,29 @@
+#include <wtf/text/WTFString.h>
+
+WebKitQueryPermissionRequest* webkitQueryPermissionRequestCreate(const WTF::String& permissionName, const WTF::String& origin, CompletionHandler<void(std::optional<WebCore::PermissionState>)>&&);
+
```
Remove this empty line

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
```diff
@@ -378,6 +379,12 @@ class UIClient : public API::UIClient {
 
+    void queryPermission(const WTF::String& permissionName, API::SecurityOrigin& origin, CompletionHandler<void(std::optional<WebCore::PermissionState>)>&& completionHandler) final
+    {
+        auto permissionRequest = adoptGRef(webkitQueryPermissionRequestCreate(permissionName, origin.securityOrigin().toString(), WTFMove(completionHandler)));
```
Better pass the origin and not its string representation and build a `WebKitSecurityOrigin` from it

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -2249,6 +2252,30 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
+     *
+     * If the signal is not handled, the query result will be set to %WEBKIT_QUERY_PERMISSION_PROMPT automatically.
+     *
+     * Returns: A #WebKitQueryPermissionResult.
```
As I said, we need to support handling this asynchronously. So, this should be a true handled signal where you can take a ref of the passed in object and return `TRUE`.

philn replied:
That contradicts the first review... I need to rewrite this another time :(

carlosgcampos replied:
I'm sorry I didn't explain it well, I said a signal that the user is expected to reply, I didn't mean it had to be replied synchronously using the callback return value.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
```diff
@@ -374,6 +375,9 @@ struct _WebKitWebViewClass {
 
+    gboolean   (* query_permission)            (WebKitWebView               *web_view,
+                                                WebKitQueryPermissionRequest *permission_request);
+
```
You need to remove a padding to not break ABI.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,153 @@
+        : permissionName(permissionName.utf8())
+        , completionHandler(WTFMove(completionHandler))
+    {
+        securityOrigin = API::SecurityOrigin::create(origin.securityOrigin());
```
Can this be moved to be initialized after conpletionHandler?

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,153 @@
+}
+
+WebKitQueryPermissionRequest*
+webkit_query_permission_request_ref(WebKitQueryPermissionRequest* queryPermissionRequest)
```
This needs to be documented

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,153 @@
+    return queryPermissionRequest;
+}
+
+void webkit_query_permission_request_unref(WebKitQueryPermissionRequest* queryPermissionRequest)
```
This needs to be documented

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,153 @@
+ * Since: 2.40
+ */
+const gchar*
+webkit_query_permission_get_name(WebKitQueryPermissionRequest* request)
```
If we kep the name this should be `webkit_query_permision_request_get_name`

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,153 @@
+
+    CString permissionName;
+    CompletionHandler<void(std::optional<WebCore::PermissionState>)> completionHandler;
+    RefPtr<API::SecurityOrigin> securityOrigin;
```
Could we keep a `WebKitSecurityOrigin` instead? that way the getter can be transfer none.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,153 @@
+ * Since: 2.40
+ */
+WebKitSecurityOrigin *
+webkit_query_permission_get_security_origin(WebKitQueryPermissionRequest* request)
```
Same here, if we keep the name, the prefix is `webkit_quety_permission_request`

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,153 @@
+ * Since: 2.40
+ */
+void
+webkit_permission_query_finish(WebKitQueryPermissionRequest* request, WebKitPermissionState state)
```
And same about the prefix

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,153 @@
+ * @request: a #WebKitQueryPermissionRequest
+ * @state: a #WebKitPermissionState
+ *
+ * Notify the web-engine of the selected permission state for the given request.
```
Add here that this should only be called as a response to the `WebKitWebView::query-permission` signal

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.cpp:
```diff
@@ -0,0 +1,153 @@
+void
+webkit_permission_query_finish(WebKitQueryPermissionRequest* request, WebKitPermissionState state)
+{
+    switch (state) {
```
Add g_return macros to check request is not null. We could also add another to check completionHandler hasn't been called already, or an early return, but it'sd actually a programmer error to call this more than once.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitQueryPermissionRequest.h.in:
```diff
@@ -0,0 +1,72 @@
+webkit_query_permission_request_ref(WebKitQueryPermissionRequest* query_permission_request);
+
+WEBKIT_API void
+webkit_query_permission_request_unref(WebKitQueryPermissionRequest* query_permission_request);
```
Move these after the `get_type` one

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -2249,6 +2252,30 @@ static void webkit_web_view_class_init(WebKitWebViewClass* webViewClass)
+     * specified by the [Permissions W3C Specification](https://w3c.github.io/permissions/).
+     *
+     * If the signal is not handled, the query result will be set to %WEBKIT_QUERY_PERMISSION_PROMPT automatically.
+     *
```
Explain here that this can be handled asynchronously by raking a ref of the query and returning `FALSE`. Check other signals.

carlosgcampos commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
```diff
@@ -2845,6 +2872,13 @@ bool webkitWebViewShowOptionMenu(WebKitWebView* webView, const IntRect& rect, We
+void webkitWebViewMakeQueryPermissionRequest(WebKitWebView* webView, WebKitQueryPermissionRequest* request)
+{
+    gboolean result;
+    // phil
```
Leftover comment here

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
```diff
@@ -960,6 +991,45 @@ static void testWebViewMediaKeySystemPermissionRequests(UIClientTest* test, gcon
+    const gchar* result = test->waitUntilQueryPermissionResultMessageReceived();
+    g_assert_cmpstr(result, ==, test->m_expectedQueryPermissionReply);
+
+    // Test allowing a permission request.
```
This is not allowing

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
```diff
@@ -253,6 +253,24 @@ class UIClientTest: public WebViewTest {
+    static gboolean queryPermissionRequested(WebKitWebView*, WebKitQueryPermissionRequest* request, UIClientTest* test)
+    {
+        WebKitPermissionState state = WEBKIT_PERMISSION_STATE_PROMPT;
+
```
`get_name` and `get_origin` are untested.

carlosgcampos reviewed (changes requested):
Looks great, thanks. We just need to decide a better name, I don't like the word request there and the documentation still says `permission request`. I think `WebKitPermissionQuery` could work, but we can be even more explicit and use `WebKitPermissionStateQuery` to make it even clearer this is about the state of the permissions, not requesting any permission. This would require to change the signal name too as `query-permission-state`

mcatanzaro commented on Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
```diff
@@ -374,8 +375,8 @@ struct _WebKitWebViewClass {
 #endif
 
-    /*< private >*/
-    void (*_webkit_reserved0) (void);
```
Nit: remove `(*_webkit_reserved5)` from the bottom of the list when you add your new vfunc at the top of the list. `(*_webkit_reserved0)` is the last to be removed.

philn replied:
But `*_webkit_reserved5` is specific to WPE.

mcatanzaro replied:
This is really not very important, but what I would do is:

```
    gboolean   (* query_permission_state)      (WebKitWebView               *web_view,
                                                WebKitPermissionStateQuery  *query);

#if PLATFORM(WPE)
    /*< private >*/
    void (*_webkit_reserved0) (void);
    void (*_webkit_reserved1) (void);
    void (*_webkit_reserved2) (void);
    void (*_webkit_reserved3) (void);
    void (*_webkit_reserved4) (void);
#endif
```

BTW I notice you dropped the `/*< private >*/`, probably by mistake?

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
```diff
@@ -253,6 +253,30 @@ class UIClientTest: public WebViewTest {
+
+        g_assert_cmpstr(webkit_permission_state_query_get_name(query), ==, "screen-wake-lock");
+
+        gchar* origin = webkit_security_origin_to_string(webkit_permission_state_query_get_security_origin(query));
```
GUniquePtr<char>

carlosgcampos commented on Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
```diff
@@ -253,6 +253,30 @@ class UIClientTest: public WebViewTest {
+    {
+        WebKitPermissionState state = WEBKIT_PERMISSION_STATE_PROMPT;
+
+        if (!g_strcmp0(test->m_expectedQueryPermissionReply, "granted"))
+            state = WEBKIT_PERMISSION_STATE_GRANTED;
+        if (!g_strcmp0(test->m_expectedQueryPermissionReply, "denied"))
+            state = WEBKIT_PERMISSION_STATE_DENIED;
```
I made a comment that I can't find, so I guess I didn't press the comment button or it was lost somehow. Anyway, I said that we should test the case of not handling the signal. I would add a new expectation, or example "prompt-default" that simply returns `FALSE` without calling the finish method.

carlosgcampos reviewed (approved):
Excellent!

philn committed:
[WPE][GTK] Enable permissions request API
https://bugs.webkit.org/show_bug.cgi?id=247065

Reviewed by Carlos Garcia Campos.

This is exposed in the GLib API as a WebKitPermissionStateQuery signalled by the WebView using the
`query-permission-state` GObject signal. When the application has a handler for this signal it has
to eventually call `webkit_permission_state_query_finish()` with a valid `WebKitPermissionState`. If
not handled, the default result is `WEBKIT_PERMISSION_STATE_PROMPT`, thus instructing the User-Agent
that it has to explicitely present a prompt asking for permission to the given feature.

* Source/WTF/Scripts/Preferences/WebPreferencesExperimental.yaml:
* Source/WebKit/PlatformGTK.cmake:
* Source/WebKit/PlatformWPE.cmake:
* Source/WebKit/SourcesGTK.txt:
* Source/WebKit/SourcesWPE.txt:
* Source/WebKit/UIProcess/API/glib/WebKitAutocleanups.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitPermissionStateQuery.cpp: Added.
(_WebKitPermissionStateQuery::_WebKitPermissionStateQuery):
(_WebKitPermissionStateQuery::~_WebKitPermissionStateQuery):
(webkitPermissionStateQueryCreate):
(webkit_permission_state_query_ref):
(webkit_permission_state_query_unref):
(webkit_permission_state_query_get_name):
(webkit_permission_state_query_get_security_origin):
(webkit_permission_state_query_finish):
* Source/WebKit/UIProcess/API/glib/WebKitPermissionStateQuery.h.in: Added.
* Source/WebKit/UIProcess/API/glib/WebKitPermissionStateQueryPrivate.h: Added.
* Source/WebKit/UIProcess/API/glib/WebKitUIClient.cpp:
* Source/WebKit/UIProcess/API/glib/WebKitWebView.cpp:
(webkit_web_view_class_init):
(webkitWebViewPermissionStateQuery):
* Source/WebKit/UIProcess/API/glib/WebKitWebView.h.in:
* Source/WebKit/UIProcess/API/glib/WebKitWebViewPrivate.h:
* Source/WebKit/UIProcess/API/gtk/webkit.h:
* Source/WebKit/UIProcess/API/gtk/webkit2.h:
* Source/WebKit/UIProcess/API/wpe/webkit.h:
* Tools/TestWebKitAPI/Tests/WebKitGLib/TestUIClient.cpp:
(testWebViewQueryPermissionRequests):
(beforeAll):

Canonical link: https://commits.webkit.org/257787@main
